stability: Flush batch iterator before computing stats; put txn record on LHS of split

spencerkimball opened the pull request:
When splitting, it makes more sense to locate the txn record on the LHS of
the split so that it can be GC'd properly in `intentResolve.processIntentsAsync`.
Previously, it would fail every time because it was added as a write command
to the LHS and they txn key was on the RHS.

Fixes #7652

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/7903)

<!-- Reviewable:end -->

spencerkimball committed:
Flush batch iterator before computing stats; put txn record on LHS of split

When splitting, it makes more sense to locate the txn record on the LHS of
the split so that it can be GC'd properly in `intentResolve.processIntentsAsync`.
Previously, it would fail every time because it was added as a write command
to the LHS and they txn key was on the RHS.

Fixes #7652

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

---

Review status: 0 of 6 files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

*[storage/replica_command.go, line 2203 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-VjzE0eBrb-FNLatj:-KN-VjzF5eUUVszelWYM:-1710434977) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2203)):*
> ```Go
> 	log.Trace(ctx, "found split key")
> 
> 	// Create new range descriptor with newly-allocated Range ID.
> ```

`s/with/with the/g`

---

*[storage/replica_command.go, line 2218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-Vn1uMk2plGcOH-yX:-KN-Vn1uMk2plGcOH-yY:-114396222) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2218)):*
> ```Go
> 		log.Trace(ctx, "split closure begins")
> 		defer log.Trace(ctx, "split closure ends")
> 		// Update existing range descriptor for first half of split. Note
> ```

I dislike the `first half` and `second half` terminology. `LHS` and `RHS` feel much more natural to my western mind. Mildly confusing to use both `first half` and `left hand side` in the same paragraph. 

Not for this PR, but it would be nice to do a cleanup of this code and replace `updated` with `lhs` or `left` and `new` with `rhs` or `right`.

---

*[storage/replica_command.go, line 2226 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-W0T4yOBrAesPYYdf:-KN-W0T4yOBrAesPYYdg:-276273621) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2226)):*
> ```Go
> 			return err
> 		}
> 		// Create range descriptor for second half of split.
> ```

`s/second half/right hand side/g`

---

*[storage/replica_command.go, line 2227 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-Wj2LM7UhY9e-QZdY:-KN-Wj2MNC0PrlAhj4cB:-533365637) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2227)):*
> ```Go
> 		}
> 		// Create range descriptor for second half of split.
> 		desc2Key := keys.RangeDescriptorKey(newDesc.StartKey)
> ```

See above. What is `desc2Key` in relation to `newDesc`? 

---

*[storage/engine/rocksdb.go, line 755 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-XZRpc0wqv73tfjEl:-KN-XZRpc0wqv73tfjEm:952485593) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/engine/rocksdb.go#L755)):*
> ```Go
> }
> 
> func (r *rocksDBBatchIterator) ComputeStats(start, end MVCCKey, nowNanos int64) (enginepb.MVCCStats, error) {
> ```

How was this not causing problems before the rest of this PR?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-:-KN-X8LH28N-_5Oipwla:980077423)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[storage/intent_resolver.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN0A-LnKb1i3Ickh30l:-KN0A-LnKb1i3Ickh30m:1513134072) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/intent_resolver.go#L334)):*
> ```Go
> 				gcArgs := roachpb.GCRequest{
> 					Span: roachpb.Span{
> 						Key: txn.Key,
> ```

Does this work? It looks to me like `kv.truncate()` mishandles requests with the `isRange` flag that don't have an end key set. We need a test that this record is actually getting GC'd. 

---

*[storage/replica_command.go, line 2218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-Vn1uMk2plGcOH-yX:-KN0Auw4qkSUIEj3uvuR:2104037204) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2218)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> I dislike the `first half` and `second half` terminology. `LHS` and `RHS` feel much more natural to my western mind. Mildly confusing to use both `first half` and `left hand side` in the same paragraph. 
> 
> Not for this PR, but it would be nice to do a cleanup of this code and replace `updated` with `lhs` or `left` and `new` with `rhs` or `right`.
</details>
+1. The "updated" and "new" terminology (and `desc{1,2}Key`) has been driving me crazy as I've been looking at this code.

---

*[storage/engine/rocksdb.go, line 755 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-XZRpc0wqv73tfjEl:-KN0DS72e6OgGoSk6jqW:-591740995) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/engine/rocksdb.go#L755)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> How was this not causing problems before the rest of this PR?
</details>
By accident. We have two sources of stats: there's an `ms *MVCCStats` object that gets passed around and contains all the pending stats (even if they're unflushed), and for the unique case of the left side of a split we compute the stats directly from the engine. As it happens, `EndTransaction` flushes its batch several times. The only thing that happens between the call to ComputeStats and the prior flush is writing to the transaction record (to persist the external intents). When the transaction record was on the RHS, everything was fine because everything that mattered for the LHS was already flushed (and the RHS stats came from the in-memory `ms` so flushes didn't matter). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-:-KN0CBqA5AQcaRGHj-fj:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/intent_resolver.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN0A-LnKb1i3Ickh30l:-KN1HE0dZGxTgPQU7tlH:329700165) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/intent_resolver.go#L334)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> Does this work? It looks to me like `kv.truncate()` mishandles requests with the `isRange` flag that don't have an end key set. We need a test that this record is actually getting GC'd. 
</details>
Yeah, this doesn't look good (even if it works, it's incorrect). As long as GCRequest is a range request, we must give it a valid range. Just give it `txn.Key` to `txn.Key.Next()`.

We *should* have a test for this. If not, well, we really need one.

---

*[storage/replica_command.go, line 2220 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN1HekUQkQyWJ-i4IS-:-KN1HekUQkQyWJ-i4IS0:1427986639) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2220)):*
> ```Go
> 		// Update existing range descriptor for first half of split. Note
> 		// that we mutate the descriptor for the left hand side of the
> 		// split first to locate the txn record there..
> ```

double dot.

---

*[storage/replica_command.go, line 2490 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN1Hn-kv7qE7hDnWQNY:-KN1Hn-kv7qE7hDnWQNZ:-880786207) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2490)):*
> ```Go
> 		// This is a stop-gap to achieve this: If there is an uninitialized
> 		// Replica, disable it now (but leave it in the map). The split will
> 		// then delete it after we are committed.
> ```

Probably want to remove this edit since I just merged a PR which takes this out (so you're saving a rebase conflict by removing your edit from your WIP).

---

*[storage/store.go, line 1554 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN1IGmCO1cGCMV9k7cY:-KN1IGmCO1cGCMV9k7cZ:1843539321) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/store.go#L1554)):*
> // It allocates a new
> // range ID and copies the supplied replicas slice, renumbering
> // replicas with indexes 1 to len(replicas).

It allocates a new RangeID and returns a RangeDescriptor whose Replicas are a copy of the supplied Replicas slice, with appropriate ReplicaIDs assigned.

---

*[storage/engine/rocksdb.go, line 755 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-XZRpc0wqv73tfjEl:-KN1Izho81SvsdWRRbrD:24150031) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/engine/rocksdb.go#L755)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> By accident. We have two sources of stats: there's an `ms *MVCCStats` object that gets passed around and contains all the pending stats (even if they're unflushed), and for the unique case of the left side of a split we compute the stats directly from the engine. As it happens, `EndTransaction` flushes its batch several times. The only thing that happens between the call to ComputeStats and the prior flush is writing to the transaction record (to persist the external intents). When the transaction record was on the RHS, everything was fine because everything that mattered for the LHS was already flushed (and the RHS stats came from the in-memory `ms` so flushes didn't matter). 
</details>
Oh wow, that was a subtle bug. Definitely not the stats problem I expected.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-:-KN1J5R02JB1hTxP_G0y:1417088721)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/replica_command.go, line 2218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-Vn1uMk2plGcOH-yX:-KN1J8TcCmGYCAdkZ6Cw:5172524) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2218)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> +1. The "updated" and "new" terminology (and `desc{1,2}Key`) has been driving me crazy as I've been looking at this code.
</details>
+2

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[storage/intent_resolver.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN0A-LnKb1i3Ickh30l:-KN2bs99ArHA8pCSjto2:-538592519) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/intent_resolver.go#L334)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Yeah, this doesn't look good (even if it works, it's incorrect). As long as GCRequest is a range request, we must give it a valid range. Just give it `txn.Key` to `txn.Key.Next()`.
> 
> We *should* have a test for this. If not, well, we really need one.
</details>
I added `txn.Key.Next()`.

---

*[storage/replica_command.go, line 2218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-Vn1uMk2plGcOH-yX:-KN2e7VsOSaUN-3GhA9N:262108080) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2218)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> +2
</details>
Fixed a lot of the stuff but left TODOs to change the protobufs and triggers.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/intent_resolver.go, line 334 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN0A-LnKb1i3Ickh30l:-KN2hLLSrTh2N-pJR0l8:1774736299) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/intent_resolver.go#L334)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> I added `txn.Key.Next()`.
</details>
`roachpb.Key(txn.Key).Next()`.

---

*[storage/replica_command.go, line 2144 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN2hQjLIvh_pCgM9OHm:-KN2hQjLIvh_pCgM9OHn:-1607804166) ([raw file](https://github.com/cockroachdb/cockroach/blob/fe5dfa66eabd456436dec0b4f923e823455c3aca/storage/replica_command.go#L2144)):*
> ```Go
> // args.SplitKey (if provided) or an internally computed key that aims
> // to roughly equipartition the range by size. The split is done
> // inside of a distributed txn which writes updated left and and new
> ```

s/and and/and/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Still needs a test that the intents for the split transaction are getting cleaned up correctly.

---

Review status: 4 of 6 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/replica_command.go, line 2218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-Vn1uMk2plGcOH-yX:-KN2jneH9s9oc7wJhFaW:-1912936449) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2218)):*
<details><summary><i>Previously, spencerkimball (Spencer Kimball) wrote…</i></summary>
> Fixed a lot of the stuff but left TODOs to change the protobufs and triggers.
</details>
Are you going to follow up with a PR to change those? If not, let's file an issue. TODOs tend to get forgotten. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-:-KN2jwijfUYndBARxuns:267213292)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Reviewed 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 6 of 7 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[storage/replica_command.go, line 2218 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN-Vn1uMk2plGcOH-yX:-KN2kGbkpKU3AAl42Aey:1330204583) ([raw file](https://github.com/cockroachdb/cockroach/blob/4d92e8b25223a0484803d5034b8078513d4ceaa6/storage/replica_command.go#L2218)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Are you going to follow up with a PR to change those? If not, let's file an issue. TODOs tend to get forgotten. 
</details>
Yes that's the plan.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 6 of 7 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[storage/client_split_test.go, line 238 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN3Wj5wKRSwPtw31xaL:-KN3Wj5wKRSwPtw31xaM:1421871653) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5f41662c3dc34f3bd6772433be8b91dba451a0f/storage/client_split_test.go#L238)):*
> ```Go
> 			t.Errorf("unexpected system key: %s; txn record should have been cleaned up", iter.Key())
> 		}
> 	}
> ```

LGTM, though I think you need to look for unexpected range descriptor keys on the right-hand side of the split as well.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903)*
<!-- Sent from Reviewable.io -->

spencerkimball commented:
Review status: 6 of 7 files reviewed at latest revision, 11 unresolved discussions, some commit checks pending.

---

*[storage/client_split_test.go, line 238 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN3Wj5wKRSwPtw31xaL:-KN3Y8vmAcvunzEf1551:24318887) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5f41662c3dc34f3bd6772433be8b91dba451a0f/storage/client_split_test.go#L238)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> LGTM, though I think you need to look for unexpected range descriptor keys on the right-hand side of the split as well.
</details>
Yep, I changed it to check both sides of the split.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

---

Review status: 6 of 7 files reviewed at latest revision, 12 unresolved discussions, some commit checks pending.

---

*[storage/client_split_test.go, line 225 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-KN3Xfh4pt2dpE7u0veF:-KN3Xfh4pt2dpE7u0veG:1358935802) ([raw file](https://github.com/cockroachdb/cockroach/blob/e5f41662c3dc34f3bd6772433be8b91dba451a0f/storage/client_split_test.go#L225)):*
> ```Go
> 	for _, key := range []roachpb.Key{keys.RangeDescriptorKey(roachpb.RKeyMin), keys.RangeDescriptorKey(splitKeyAddr)} {
> 		if _, _, err := engine.MVCCGet(context.Background(), store.Engine(), key, store.Clock().Now(), true, nil); err != nil {
> 			t.Errorf("failed to read consistent range descriptor (intent still exists?) for key %s: %s", key, err)
> ```

Don't say "intent still exists?" unless you check that `err` is a `WriteIntentError`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/7903#-:-KN3Yo3DS5AX1PcYkRoY:676834233)*
<!-- Sent from Reviewable.io -->
